src/aspire/estimation/covar2d.py

```diff
@@ -9,25 +9,107 @@
     J. Struct. Biol. 195, 27-81 (2016). DOI: 10.1016/j.jsb.2016.04.013
     """
 
+    def __init__(self, basis, as_type='single'):
+        """
+        constructor of an object for 2D covariance analysis
+        """
+        self.basis = basis
+        ensure(basis.d == 2, 'Only two-dimensional basis functions are needed.')
+        self.as_type = as_type
+
+    def get_mean(self, coeffs):
+        """
+        Calculate the mean vector from the expansion coefficients of 2D images without CTF information.
+
+        :param coeffs: A coefficient vector (or an array of coefficient vectors) to be averaged.
+        :return: The mean value vector for all images.
+        """
+        if coeffs is None:
+            raise RuntimeError('The coefficients need to be calculated first!')
+
+        mask = self.basis._indices["ells"] == 0
+        mean_coeff = np.zeros((self.basis.basis_count, 1), dtype=self.as_type)
+        mean_coeff[mask, 0] = np.mean(coeffs[mask, ...], axis=1)
+
+        return mean_coeff
+
+    def get_covar(self, coeffs, mean_coeff=None,  do_refl=True):
+        """
+        Calculate the covariance matrix from the expansion coefficients without CTF information.
+
+        :param coeffs: A coefficient vector (or an array of coefficient vectors) calculated from 2D images.
+        :param mean_coeff: The mean vector calculated from the `coeffs`.
+        :param do_refl: If true, enforce invariance to reflection (default false).
+        :return: The covariance matrix of coefficients for all images.
+        """
+        if coeffs is None:
+            raise RuntimeError('The coefficients need to be calculated first!')
+        if mean_coeff is None:
+            mean_coeff = self.get_mean(coeffs)
+
+        covar_coeff = []
+        ind = 0
+        ell = 0
+        mask =  self.basis._indices["ells"] == ell
+        coeff_ell = coeffs[mask, ...] - mean_coeff[mask, ...]
+        covar_ell = np.array(coeff_ell @ coeff_ell.T/np.size(coeffs, 1))
+        covar_coeff.append(covar_ell)
+        ind += 1
+
+        for ell in range(1, self.basis.ell_max+1):
+            mask = self.basis._indices["ells"] == ell
+            mask_pos = [mask[i] and (self.basis._indices['sgns'][i] == +1) for i in range(len(mask))]
+            mask_neg = [mask[i] and (self.basis._indices['sgns'][i] == -1) for i in range(len(mask))]
+            covar_ell_diag = np.array(coeffs[mask_pos, :] @ coeffs[mask_pos, :].T +
+                coeffs[mask_neg, :] @ coeffs[mask_neg, :].T) / (2 * np.size(coeffs, 1))
+
+            if do_refl:
+                covar_coeff.append(covar_ell_diag)
+                covar_coeff.append(covar_ell_diag)
+                ind = ind+2
+            else:
+                covar_ell_off = np.array((coeffs[mask_pos, :] @ coeffs[mask_neg, :].T / np.size(coeffs, 1) -
+                                 coeffs[mask_neg, :] @ coeffs[mask_pos, :].T)/(2*np.size(coeffs, 1)))
```

Comment:
Space around division sign and asterisk.

Reply:
done.